### DISCUSSION TITLE
Per-position entropy calculation using MSA-ESM-1b w/ MSA

### DISCUSSION THREAD

**Discussion by YaoYinYing** [facebookresearch/esm#295] [Q&A]
Hi all,

Thanks for the project!

I read the ESM-1v paper and it says that MSA-ESM-1b outperforms all the rest models which use a single sequence as input, and the entropy value of per-position refers to functional or non-functional site(s).

I want ask whether the entropy calculated with MSA info via MSA-ESM-1b makes sense, and if so, how can we make this avaliable? I will be appreciated if any one of you can provide some example codes.

Thank you very much!

**Comment by tomsercu** [comment]
Sorry I don't understand this question.  What is the problem you want to solve? What is the model you're trying to use: ESM-1b or MSA Transformer?

**Reply by YaoYinYing** [reply]
I'm trying to label all the designable positions of an enzyme for themostability improvement while leave the entire catalytic region alone. 

According to the ESM-1V paper, the higher per-position entropy refers to solvent accessible surface in protien.

I'd like to use MSA transformer to minimize the number of positions for design and reserving more phylogenetic property of the protein family.

So my question is, since the entropy reflects the molecule basis in protein, does it make sense to use MSA transformer for entropy calculation?

**Reply by adrienchaton** [reply]
it seems that there are some intuitions based on biology and observations in the protein language model literature supporting what you are asking @YaoYinYing 
_ low entropy residues tend to be functional sites
_ high entropy residues tend to be surface residues (which goes along the inverse folding paper showing higher LM perplexity on surface residues)
whether it makes more sense to calculate entropy with ESM1b or MSA transformer, I guess there is no clear answer than try
and you can simply use torch.distributions.categorical.Categorical(logits).entropy() on the logits output of the LM
